sirepo/auth_role_moderation.py

```diff
@@ -0,0 +1,116 @@
+
+    def _dns_reverse_lookup(ip):
+        import dns.resolver, dns.reversename
+        try:
+            return str(dns.resolver.query(dns.reversename.from_address(ip), 'PTR')[0])
+        except (dns.resolver.NoAnswer, dns.resolver.NXDOMAIN):
+            return 'No Reverse DNS Lookup'
+
+    def _send_moderation_email(info):
+        sirepo.smtp.send(
+            recipient=_cfg.moderator_email,
+            subject=f'{info.sim_type} Access Request',
+            body=pkjinja.render_resource('auth_role_moderation/moderation_email', info),
+        )
+
+    req = sirepo.http_request.parse_post(require_sim_type=False)
+    d = req.req_data
+    uid = sirepo.auth.logged_in_user()
+
+    sirepo.auth_db.UserRoleInvite(
+        uid=uid,
+        role=sirepo.auth_role.for_sim_type(d.simulationType),
+        status='pending',
+        token=sirepo.util.random_base62(32),
+    ).save()
+    l = _moderation_link(d.simulationType)
+    if sirepo.auth.dev_and_no_method('email'):
+        pkdlog('Email not enabled. Moderation reason={} Link to approve/deny={}', d.reason, l)
+    else:
+        _send_moderation_email(
+            PKDict(
+                display_name=sirepo.auth_db.UserRegistration.search_by(uid=uid).display_name,
+                email_addr=sirepo.auth.get_module('email').AuthEmailUser.search_by(uid=uid).unverified_email,
+                ip_addr=flask.request.remote_addr,
+                domain_name=_dns_reverse_lookup(flask.request.remote_addr),
+                link=l,
+                reason=d.reason,
+                role=sirepo.auth_role.for_sim_type(d.simulationType),
+                sim_type=d.simulationType,
+                uid=uid,
+                user_agent=flask.request.headers.get("User-Agent"),
+            )
+        )
+    return sirepo.http_reply.gen_json_ok()
+
+
+@sirepo.api_perm.require_adm
+def api_getPendingUserRoleInvites():
+    return sirepo.http_reply.gen_json(PKDict(
+        rows=[r.as_pkdict() for r in sirepo.auth_db.UserRoleInvite.get_pending_rows()],
+        headers=[c for c in sirepo.auth_db.UserRoleInvite.column_names() if c != 'token'],
+    ))
+
+
+@sirepo.api_perm.require_adm
+def api_approveOrDenyRole():
+    def _send_approve_or_deny_email(info):
+        a = 'Approved' if info.approved else 'Denied'
+        if sirepo.auth.dev_and_no_method('email'):
+            pkdlog('Email not enabled. Access is {}', a)
+            return
+        sirepo.smtp.send(
+            recipient=sirepo.auth.get_module('email').AuthEmailUser.search_by(uid=info.uid).unverified_email,
+            subject=f'{info.role} Access Request' + a,
+            body=pkjinja.render_resource('auth_role_moderation/approve_email' if info.approved else 'auth_role_moderation/deny_email', PKDict(
+                link=flask.url_for('_dispatch_empty', _external=True) + sirepo.auth_role.sim_type(info.role),
+                app_name=sirepo.simulation_db.SCHEMA_COMMON.appInfo[sirepo.auth_role.sim_type(info.role)].shortName,
```

Comment:
Needs to be abstracted. auth_role should be able to go both directions.

Reply:
> app_name=sirepo.simulation_db.SCHEMA_COMMON.appInfo[sirepo.auth_role.sim_type(info.role)].shortName

I just think the abstraction of looking up appInfo to get the shortName should be somewhere and not inline in auth_role_moderation. It will be needed in the list of moderation requests as well.

> link=flask.url_for('_dispatch_empty', _external=True) + sirepo.auth_role.sim_type(info.role),

Similarly this should be encapsulated in sirepo.uri_router. This is a bit complicated, because sirepo.uri.app_root already defines the concept of mapping from a sim_type to uri, but it doesn't support absolute position. Attributes that begin with an underscore (_dispatch_empty) should not be referenced outside the module in which they are defined. This is called a leaky abstraction. I think the best place to do this would be in sirepo.uri_router so this would look like (putting it all together:

```py
t = sirepo.auth_role.sim_type(info.role)
m = 'approve' if info.approved else 'deny'
body=pkjinja.render_resource(
    f'auth_role_moderation/{m}_email',
    PKDict(
        link=sirepo.uri_router.absolute_uri(sirepo.uri.app_root(t)),
        app_name=sirepo.simulation_db.app_info(t).shortName,
```

At least that's how I see. It could be that app_info is in the wrong place.

Reply:
I had already made some changes to this section at @e-carlin's recommendation, including moving app_name out to the main function body and adding sirepo.uri.server_root() to be used in the link. So now I'm a bit confused on how to implement these changes- can you please let me know if/how they still apply?

Reply:
7813ef1ac1103a4c5aca70b302c249a7d56d910f